torch/linalg/__init__.py

```diff
@@ -213,6 +213,96 @@
             [-3.1113,  2.7381]], dtype=torch.float64)
 """)
 
+lstsq = _add_docstr(_linalg.linalg_lstsq, r"""
+torch.linalg.lstsq(input, b, cond=None, driver_name=None) -> (Tensor x, Tensor rank, Tensor s)
+
+Computes the least squared solution to the system with a batch of matrices :math:`a` (represented by :attr:`input`)
+and a batch of vectors/matrices :math:`b` such that
+
+.. math::
+    x = \text{argmin}_x \|ax - b\|_F,
+
+where :math:`a` is of size :math:`(..., m, n)` and :math:`b` is of size :math:`(..., m, k)` or
+:math:`(..., m)`. The batch dimensions of :math:`a` and :math:`b` have to be broadcastable.
+
+.. note::
+    The input matrices/vectors are supposed to be non-empty, i.e. :math:`m > 0, m > 0, k > 0`.
+    The empty batch dimensions are allowed however.
+
+The returned solution :math:`x` is of shape :math:`(..., \max(m, n), k)` if :math:`b` is of shape :math:`(..., m, k)`,
+and is of shape :math:`(..., \max(m, n), 1)` if :math:`b` is of shape :math:`(..., m)`.
+The batch sizes of :math:`x` is the broadcasted shape of the batch dimensions of :math:`a` and :math:`b`.
+A minimizer (not necessarily unique for rank-deficient matrices)
+could be obtained by slicing the solution with, for example, ``x.narrow(-2, 0, n)``.
+The remaining entries ``x.narrow(-2, n, max(m, n) - n)``
+encode the residual sum of squares for the solution in each column for full-rank matrices in :math:`a`,
+i.e. if :math:`\text{rank}(a[..., :, :]) = \min(m, n)`, the residuals for the system :math:`a[..., :, :]` could be obtained by
+``residuals = x.narrow(-2, n, max(m, n) - n).pow(2).sum(-2)[..., :, :]``.
+
+.. note::
+    The case when :math:`m < n` is not supported on the GPU yet.
+
```

Comment:
Turns out cuSolver does handle the case `m < n`. Does it make sense to enable cuSolver gels version then?
The difference with MAGMA is that instead of returning a solution of shape `(max(m, n), nrhs)`, it will return a solution of shape `(n, nrhs)` instead.

Reply:
Hm, and there is cublas routine as well. I should have done my research better instead of just concentrating on MAGMA...

Reply:
In your CuPy examples you are setting to zero columns of `b` while, I think, you wanted to set to zero columns of `a` to decrease its rank. Both cuBLAS and cuSOLVER don't handle rank-deficient case, and both have the requirement that `m >= n`.

Reply:
Correct, confused the order, it is different between the libraries.